src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -124,22 +129,95 @@ internal void AppendMetadata()
 
                 TypeDefinitionHandle typeDefinitionHandle = AddTypeDefinition(typeBuilder, parent, _nextMethodDefRowId, _nextFieldDefRowId);
                 Debug.Assert(typeBuilder._handle.Equals(typeDefinitionHandle));
+                WriteCustomAttributes(typeBuilder._customAttributes, typeDefinitionHandle);
 
-                // Add each method definition to metadata table.
-                foreach (MethodBuilderImpl method in typeBuilder._methodDefStore)
+                foreach (MethodBuilderImpl method in typeBuilder._methodDefinitions)
                 {
-                    AddMethodDefinition(method, method.GetMethodSignatureBlob());
+                    MethodDefinitionHandle methodHandle = AddMethodDefinition(method, method.GetMethodSignatureBlob());
+                    WriteCustomAttributes(method._customAttributes, methodHandle);
                     _nextMethodDefRowId++;
+
+                    if (method._dllImportData != null)
+                    {
+                        AddMethodImport(methodHandle, method._dllImportData.EntryPoint ?? method.Name,
+                            method._dllImportData.Flags, GetModuleReference(method._dllImportData.ModuleName));
+                    }
                 }
 
-                foreach (FieldBuilderImpl field in typeBuilder._fieldDefStore)
+                foreach (FieldBuilderImpl field in typeBuilder._fieldDefinitions)
                 {
-                    AddFieldDefinition(field, MetadataSignatureHelper.FieldSignatureEncoder(field.FieldType, this));
+                    FieldDefinitionHandle fieldHandle = AddFieldDefinition(field, MetadataSignatureHelper.FieldSignatureEncoder(field.FieldType, this));
+                    WriteCustomAttributes(field._customAttributes, fieldHandle);
                     _nextFieldDefRowId++;
+
+                    if (field._offset > 0 && (typeBuilder.Attributes & TypeAttributes.ExplicitLayout) != 0)
+                    {
+                        AddFieldLayout(fieldHandle, field._offset);
+                    }
                 }
             }
         }
 
+        private ModuleReferenceHandle GetModuleReference(string moduleName)
+        {
+            _moduleReferences ??= new();
```

Comment:
```suggestion
            _moduleReferences ??= new Dictionary<string, ModuleReferenceHandle>();
```
Nit: Coding conventions - the type is not obvious so it should be explicit.